### ISSUE TITLE
Minimize amount of redirects for browsers

### ISSUE THREAD

**Issue by mrhyde** [LotusTM/Kotsu#170]
Since the way we achieve clean and nice looking URL's is based on how nginx and CloudFlare handle routing we should consider optimizing this process.
Currently, we store all content in index.html pages inside of named folder that we are using as URI but accessing this resources causes a browser to do **3** requests before we can get page content!
1. Check if `https://example.com/site/url` exists (it's not)
2. Follow to new resource location that was specified in response header of the first request, `http://example.com/site/url/` (note trailing slash and HTTP protocol)
3. Follow to HTTPS location of the same page (CloudFlare rewrite)

Possible solutions:
1. Configure our nginx server to use HTTPS (probably using `Let's Encrypt`). This will cut one redirect.
2. Always use trailing slash in links to avoid another redirect

P.S. Just to specify that we are talking 10-100ms of delays for page render so it's not a critical issue.

**Comment by ArmorDarks** [comment]
I think we should make Kotsu agnostic about server environment (be it nginx or apache) as much as possible (yeap, despite we're actually launching our docker container with nginx), and it should involve minimum of server configuration, whenever possible.

Using trailing slash everywhere doesn't seem to be hard and we can update our current urls and recommend it, but there is no way to enforce them on user. Well, we actually can use cheerio to replace urls in generated files, but it seems as very prejudiced and fragile solution...

From other perspective, we can also change configuration of Kotsu's nginx. If someone will transfer to other server solution and won't implement same behavior, at least nothing will break, so I think it should be ok.

**Comment by ArmorDarks** [comment]
👍

**Comment by ArmorDarks** [comment]
As unexpected side effect, it also fixes issue with wrongly encoded non-latin urls in Microsoft Edge 14 and below.

Some quick regex to find all hrefs with no trailing slash:

```
href=(['])([^\s]+)(?<!\/)\1
```

and to replace:

```
href='$2/'
```

**Comment by mrhyde** [comment]
As per discussed during the last meeting we should consider running optional test suite to check links in the build folder.
`cheerio` is probably the best way to deal with this

**Comment by ArmorDarks** [comment]
Yeah, the problem is that we actually have to run it after the build, when test stage already passed.

We can trigger another script after `npm run build` finished, but I'm not sure — won't it be confusing?

**Comment by mrhyde** [comment]
During `npm test` we can only run unit and integration tests, functional tests will require built application. I don't see how it can be done differently.

**Comment by ArmorDarks** [comment]
I found some solutions to do this:

* https://github.com/DavidAnson/grunt-check-pages
* https://github.com/lnyarl/grunt-deadlink
* https://github.com/ChrisWren/grunt-link-checker

I guess I should check for not Grunt-based solutions too, there might be something better. However, in all cases we have to crawl launched website, like on `localhost`. I don't know would it be possible to do in CI? It is logical, because otherwise you can pick up url with `cheerio `, but against what will you check it? Against filesystem?

Besides, there is one serious issue with that. It's uneasy to predict whether url should have trailing slash or no. At first I assumed that we simply can check for dot in path. If it has dot, then it has link to file, and we can omit trailing slash. Right?

Unfortunately, nope. We can deal with `/some/url/1.0/` or even `this/is/page.html/url/` and those will be perfectly valid urls.

**Comment by ArmorDarks** [comment]
I wonder, if we will add SSL to our Nginx container, will it work for everyone, or it should be configured per each domain?

In later case, this should be taken into separate issue as task to write guideline for configuring SSL with Kotsu.

**Comment by ArmorDarks** [comment]
Documented recommended url format and specificity of our Nginx server behavior here: https://github.com/LotusTM/Kotsu/wiki/Urls#never-with-trailing-slash
